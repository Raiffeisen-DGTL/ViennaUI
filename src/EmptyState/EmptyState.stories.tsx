import React from 'react';
import { Story, Meta } from 'storybook';
import { EmptyState, EmptyStateProps } from './EmptyState';
import { Button } from '../Button';

export default {
    title: 'Development/EmptyState',
    component: EmptyState,
} as Meta;

export const Overview: Story<EmptyStateProps> = (args) => {
    return (
        <EmptyState {...args}>
            <EmptyState.Title>Оформите карту</EmptyState.Title>
            <EmptyState.Description>
                Чтобы получить выгодный кэшбэк, больше не нужно расплачиваться за бензин одной картой, а за авиабилеты —
                другой. С картой #всёсразу кэшбэк начисляется за любые покупки.
            </EmptyState.Description>
            <EmptyState.Actions>
                <Button design='accent'>Оформить</Button>
                <Button design='ghost'>Позвонить</Button>
            </EmptyState.Actions>
        </EmptyState>
    );
};
