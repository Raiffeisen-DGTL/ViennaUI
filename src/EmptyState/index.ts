export { EmptyState, type EmptyStateProps, type EmptyStateAlignment, type EmptyStateDirection } from './EmptyState';
