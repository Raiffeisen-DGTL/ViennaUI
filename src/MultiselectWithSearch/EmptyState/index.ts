export { EmptyState, type EmptyStateProps } from './EmptyState';
