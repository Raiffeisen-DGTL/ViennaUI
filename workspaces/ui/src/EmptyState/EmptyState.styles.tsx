import styled from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const presets = getPresets('emptyState', {
    base: null,
    title: null,
    description: null,
    row: null,
    content: null,
});

export const Box = styled.div`
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;

    ${presets.base}
`;

export const Title = styled.div`
    ${presets.title}
`;

export const Description = styled.div`
    ${presets.description}
`;

export const Actions = styled.div`
    display: flex;
    flex-direction: column;
`;

export const Row = styled.div`
    display: flex;
    flex-direction: column;
`;

export const Content = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    position: relative;

    ${presets.content}

    > ${Row}:not(:last-child) {
        ${presets.row}
    }
`;
