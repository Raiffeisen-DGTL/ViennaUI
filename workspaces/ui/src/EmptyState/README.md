# EmptyState

Компонент "Пустое состояние".
Компонент предназначен для использования в таких случаях, когда: результаты поиска отсутствуют, нет контента, ошибка загрузки, состояние загрузки.
Компонент также можно использовать для отображения состояния успешного выполнения действий, если не предусмотрен уникальный success page.


## Импорт

```
import { EmptyState } from 'vienna-ui';
```

## Свойства / Props

Prop | Type | Default | Description
--- | --- | --- | ---
loading | boolean \| undefined | false | Состояние загрузки. true, чтобы отобразить spinner

## Использование

Содержимое компонента передается как дочерние элементы, которые определены
`EmptyState.Title`, `EmptyState.Description`, `EmptyState.Actions` - заголовок, контент сообщения, действия для пользователя соответсвенно.
Компонент заполняет все пространство в родительском контейнере и позиционируется по центру (вертикально и горизонтально).

```jsx
<EmptyState>
    <EmptyState.Title>Оформите дебетовую Кэшбэк карту</EmptyState.Title>
    <EmptyState.Description>
        <ul>
            <li>1,5% кэшбэк рублями на все покупки</li>
            <li>Бесплатное обслуживание</li>
            <li>До 4% на остаток на счете</li>
        </ul>
    </EmptyState.Description>
    <EmptyState.Actions>
        <Button design='accent'>Оформить</Button>
        <Button design='ghost'>Подробнее</Button>
    </EmptyState.Actions>
</EmptyState>
```

## Состояние загрузки
##### Свойство `loading`

Для отображения состояния загрузки нужно задать параметр `loading`

```jsx
<EmptyState loading>
    <EmptyState.Title>Оформите карту</EmptyState.Title>
    <EmptyState.Description>
        Чтобы получить выгодный кэшбэк, больше не нужно расплачиваться за бензин одной картой, а за авиабилеты —
        другой. С картой #всёсразу кэшбэк начисляется за любые покупки.
    </EmptyState.Description>
</EmptyState>
```

## Дополнительный контент

Чтобы добавить дополнительный элемент в компонент "Пустое состояние", который не противоречит принципам и стилистике дизайн системы, его можно также передать как дочерний элемент.

```jsx
<EmptyState>
    <Logo size='xl' />
    <EmptyState.Title>Оформите карту</EmptyState.Title>
    <EmptyState.Description>
        Чтобы получить выгодный кэшбэк, больше не нужно расплачиваться за бензин одной картой, а за авиабилеты —
        другой. С картой #всёсразу кэшбэк начисляется за любые покупки.
    </EmptyState.Description>
    <EmptyState.Actions>
        <Button design='accent'>Перезагрузить</Button>
        <Button design='ghost'>Позвонить</Button>
    </EmptyState.Actions>
</EmptyState>
```
