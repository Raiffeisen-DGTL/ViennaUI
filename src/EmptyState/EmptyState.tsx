import React, { useMemo, FC } from 'react';
import { Spinner, SpinnerProps } from '../Spinner';
import { Box, Content, Row, Actions, Title, Description } from './EmptyState.styles';
import { BoxStyled } from '../Utils/styled';

export interface EmptyStateProps extends BoxStyled<typeof Box, {}> {
    /** Состояние загрузки. true, если отображать spinner */
    loading?: boolean;
    spinnerProps?: SpinnerProps;
}

export const EmptyState: FC<EmptyStateProps> & {
    Title: typeof Title;
    Description: typeof Description;
    Actions: typeof Actions;
} = (props) => {
    const { loading = false, spinnerProps, children, ...attrs } = props;

    const content = useMemo(() => {
        return React.Children.map(children, (item, index) => (item ? <Row key={index}>{item}</Row> : null));
    }, [children]);

    return (
        <Box {...attrs}>
            <Content>
                {loading && (
                    <Row>
                        <Spinner {...spinnerProps} />
                    </Row>
                )}
                {content}
            </Content>
        </Box>
    );
};

EmptyState.displayName = 'EmptyState';

EmptyState.Title = Title;
EmptyState.Description = Description;
EmptyState.Actions = Actions;
