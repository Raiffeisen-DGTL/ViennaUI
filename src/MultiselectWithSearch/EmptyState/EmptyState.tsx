import React, { FC, PropsWithChildren, useMemo } from 'react';
import {
    SearchIcon,
    Binoculars1Icon,
    ZooStorePawIcon,
    RainIcon,
    KeyboardIcon,
    AppleIcon,
    MicroscopeIcon,
    GlassesEyeIcon,
    WhaleIcon,
    FlowerIcon,
    BoatIcon,
    FishIcon,
    UfoIcon,
    PalmIcon,
    FaceNeutralIcon,
} from 'vienna.icons';
import { Box, BoxProps, Text } from './EmptyState.styles';
import { Breakpoints } from '../../Utils/responsiveness';

export interface EmptyStateProps<B = Breakpoints> {
    size?: BoxProps<B>['$size'];
    randomIcon?: boolean;
    hideIcon?: boolean;
}

const getRandomElement = <T,>(array: T[]) => {
    if (array.length === 0) {
        throw new Error('Array cannot be empty');
    }
    const index = Math.floor(Math.random() * array.length);
    return array[index];
};

export const EmptyState: FC<PropsWithChildren<EmptyStateProps>> = ({ size = 'm', randomIcon, hideIcon, children }) => {
    const iconSize = 24;
    const iconProps = { size: iconSize };
    const iconsArr = [
        Binoculars1Icon,
        ZooStorePawIcon,
        RainIcon,
        KeyboardIcon,
        AppleIcon,
        MicroscopeIcon,
        GlassesEyeIcon,
        WhaleIcon,
        FlowerIcon,
        BoatIcon,
        FishIcon,
        UfoIcon,
        PalmIcon,
        FaceNeutralIcon,
    ];
    const icon = useMemo(() => {
        return randomIcon ? React.createElement(getRandomElement(iconsArr), iconProps) : <SearchIcon {...iconProps} />;
    }, [randomIcon]);
    return (
        <Box $size={size}>
            {!hideIcon && icon}
            <Text $size={size}>{children}</Text>
        </Box>
    );
};
