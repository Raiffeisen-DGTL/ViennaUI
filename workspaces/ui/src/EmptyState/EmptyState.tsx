import React, { useMemo } from 'react';
import { Spinner } from '../Spinner';
import { Box, Content, Row, Actions, Title, Description } from './EmptyState.styles';

export type EmptyStateProps = Props & HTMLAttributeProps;

interface Props {
    /** Состояние загрузки. true, если отображать spinner */
    loading?: boolean;
}

interface HTMLAttributeProps {
    id?: string;
    title?: string;
    className?: string;
    role?: string;
    spellCheck?: boolean;
}

interface Children {
    Title: React.FC;
    Description: React.FC;
    Actions: React.FC;
}

export const EmptyState: React.FC<EmptyStateProps> & Children = (props): JSX.Element => {
    const { loading, children, ...attrs } = props;

    const content = useMemo(() => {
        return React.Children.map(children, (item, index) => <Row key={index}>{item}</Row>);
    }, [children]);

    return (
        <Box {...attrs}>
            <Content>
                {loading && (
                    <Row>
                        <Spinner />
                    </Row>
                )}
                {content}
            </Content>
        </Box>
    );
};

EmptyState.displayName = 'EmptyState';
EmptyState.defaultProps = {
    loading: false,
};
EmptyState.Title = Title;
EmptyState.Description = Description;
EmptyState.Actions = Actions;

export default EmptyState;
