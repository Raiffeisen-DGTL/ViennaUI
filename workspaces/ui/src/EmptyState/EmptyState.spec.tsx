import React from 'react';
import EmptyState, { EmptyStateProps } from './EmptyState';

test('EmptyState', () => {
    const snap = snapshot.render(<EmptyState />);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ loading', () => {
    const loading: EmptyStateProps['loading'][] = [true, false];

    loading.forEach((value) => {
        const snap = snapshot.render(<EmptyState loading={value} />);
        expect(snap).toMatchSnapshot(`loading-${value}`);
    });
});

test('EmptyState w/ title', () => {
    const component = (
        <EmptyState>
            <EmptyState.Title>EmptyState title</EmptyState.Title>
        </EmptyState>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ description', () => {
    const component = (
        <EmptyState>
            <EmptyState.Description>EmptyState description</EmptyState.Description>
        </EmptyState>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ actions', () => {
    const component = (
        <EmptyState>
            <EmptyState.Actions>EmptyState actions</EmptyState.Actions>
        </EmptyState>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ some child', () => {
    const component = (
        <EmptyState>
            <div>Some child</div>
        </EmptyState>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ title, descriptions, actions', () => {
    const component = (
        <EmptyState>
            <EmptyState.Title>EmptyState title</EmptyState.Title>
            <EmptyState.Description>EmptyState description</EmptyState.Description>
            <EmptyState.Actions>EmptyState actions</EmptyState.Actions>
        </EmptyState>
    );

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('EmptyState w/ some props', () => {
    const props: EmptyStateProps = { className: 'className', id: 'EmptyState', title: 'this is EmptyState' };
    const component = <EmptyState {...props} />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.shallow(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.props()[key]).toEqual(value);
    }
});
