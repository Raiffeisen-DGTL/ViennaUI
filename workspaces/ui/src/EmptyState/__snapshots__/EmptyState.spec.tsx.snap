// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  />
</div>
`;

exports[`EmptyState w/ actions 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Actions-sc-9kuou6-3 iOtENm"
      >
        EmptyState actions
      </div>
    </div>
  </div>
</div>
`;

exports[`EmptyState w/ description 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Description-sc-9kuou6-2 fbplxg"
      >
        EmptyState description
      </div>
    </div>
  </div>
</div>
`;

exports[`EmptyState w/ loading: loading-false 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  />
</div>
`;

exports[`EmptyState w/ loading: loading-true 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="Spinnerstyles__Box-moemrn-0 jPiLVW"
      >
        <svg
          class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            class="Spinnerstyles__Circle-moemrn-2 cElquL"
            color="primary"
            cx="50"
            cy="50"
            r="44"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`EmptyState w/ some child 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div>
        Some child
      </div>
    </div>
  </div>
</div>
`;

exports[`EmptyState w/ some props 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs className"
  id="EmptyState"
  title="this is EmptyState"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  />
</div>
`;

exports[`EmptyState w/ title 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Title-sc-9kuou6-1 dBDBIE"
      >
        EmptyState title
      </div>
    </div>
  </div>
</div>
`;

exports[`EmptyState w/ title, descriptions, actions 1`] = `
<div
  class="EmptyStatestyles__Box-sc-9kuou6-0 fYMFPs"
>
  <div
    class="EmptyStatestyles__Content-sc-9kuou6-5 kRkxxs"
  >
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Title-sc-9kuou6-1 dBDBIE"
      >
        EmptyState title
      </div>
    </div>
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Description-sc-9kuou6-2 fbplxg"
      >
        EmptyState description
      </div>
    </div>
    <div
      class="EmptyStatestyles__Row-sc-9kuou6-4 gAcuUG"
    >
      <div
        class="EmptyStatestyles__Actions-sc-9kuou6-3 iOtENm"
      >
        EmptyState actions
      </div>
    </div>
  </div>
</div>
`;
